### PR TITLE
Fix Issue #1855: "All Bazel commands fail when changing a Go dependency when using Bzlmod extension"

### LINKED ISSUE #1855: All Bazel commands fail when changing a Go dependency when using Bzlmod extension
### What version of gazelle are you using?
0.38.0

### What version of rules_go are you using?
0.48.0

### What version of Bazel are you using?
7.2.1

### Does this issue reproduce with the latest releases of all the above?
Yes

### What operating system and processor architecture are you using?
MacOS, Arm64

### What did you do?
Checkout  https://github.com/chenxiao0228/cobra/tree/bazel_demo

change "github.com/cpuguy83/go-md2man/v2 v2.0.4" -> "github.com/cpuguy83/go-md2man/v2 v2.0.3" in "go.mod"

run "bazel run gazelle", or any Bazel command "bazel build cobra", "bazel mod tidy"

### What did you expect to see?
Be able to run "bazel run @rules_go//go -- mod tidy" to fetch the new version and add its hash in "go.sum"
Or even better
Be able to run "bazel mod tidy" so everything is up to date

### What did you see instead?
All Bazel commands fail with "Error in fail: No sum for github.com/cpuguy83/go-md2man/v2@2.0.3 from parent-label-todo found. You may need to run: bazel run @rules_go//go -- mod tidy"

Even though Gazelle suggests "You may need to run: bazel run @rules_go//go -- mod tidy", the command fails with the same err msg.

### PR DESCRIPTION
**Pull request by Buzz-Lightyear**
**What type of PR is this?**
Bug fix

**What package or component does this PR mostly affect?**
Go bzlmod

**What does this PR do? Why is it needed?**
When trying to bump the version of a Go module in a repo that uses the `go_deps` bzlmod extension, we run into an issue wherein all Bazel commands fail with ` "Error in fail: No sum for <module>`. However the suggested fix is to run `bazel run @rules_go//go -- mod tidy` which fails with the same error, thus forcing us into a pickle.

This PR changes this to a warning instead of an error after which the suggested fix works.

**Which issues(s) does this PR fix?**

Fixes #1855

### CODE REVIEW

File: internal/bzlmod/go_deps.bzl [hunk 1]
```diff
@@ -689,7 +694,10 @@ def _get_sum_from_module(path, module, sums):
             # replacement have no sums, so we can skip this
             return None
         elif module.local_path == None:
-            fail("No sum for {}@{} from {} found. You may need to run: bazel run @rules_go//go -- mod tidy".format(path, module.raw_version, "parent-label-todo"))  #module.parent_label))
+            # When updating a dependency, its sum may not be in go.sum and we can't hard fail here
+            # since we need Bazel to tidy the module
+            print("No sum for {}@{} found, run bazel run @rules_go//go -- mod tidy to generate it".format(path, module.raw_version))  #module.parent_label))
```
Comment (fmeum) [inline internal/bzlmod/go_deps.bzl:699]:
```suggestion
            print("No sum for {}@{} found, run bazel run @rules_go//go -- mod tidy to generate it".format(path, module.raw_version))
```

Let's remove this garbled TODO. @stefanpenner

**Review by fmeum — COMMENTED** [review]
Could you also replace the generic `if version is specified, sum must also be` message in `go_repository` with one similar to the new warning (include the module path, version and command to run tidy).

File: internal/bzlmod/go_deps.bzl (hunk 1 above)
Comment (Buzz-Lightyear) [reply internal/bzlmod/go_deps.bzl]: Done

**Review by fmeum — APPROVED** [review]
Looks good to me. The go_repository still fails to fetch without a version, so this change has no security content.

@tyler-french Could you also take a look?

**Review by tyler-french — APPROVED** [review]

File: internal/go_repository.bzl [hunk 2]
```diff
@@ -188,8 +190,11 @@ def _go_repository_impl(ctx):
         for key in ("urls", "strip_prefix", "type", "sha256", "commit", "tag", "vcs", "remote"):
             if getattr(ctx.attr, key):
                 fail("cannot specify both version and %s" % key)
-        if ctx.attr.version and not ctx.attr.sum:
-            fail("if version is specified, sum must also be")
+        if not ctx.attr.sum:
+            if is_module_extension_repo:
+                fail("No sum for {}@{} found, run bazel run @rules_go//go -- mod tidy to generate it".format(ctx.attr.importpath, ctx.attr.version))
+            else:
+                fail("if version is specified, sum must also be")
```
Comment (fmeum) [inline internal/go_repository.bzl:197]: This required two more fixes: Replacing `print` with `fail` and also showing the new error message only with Bzlmod.

Comment (fmeum) [reply internal/go_repository.bzl:197]: @Buzz-Lightyear No need to take action, I applied the fixes and will merge now.

**Comment by stefanpenner** [comment]
Thank you!

**Pull request merged**